core/src/main/java/org/infinispan/persistence/support/NonBlockingStoreAdapter.java

```diff
@@ -0,0 +1,340 @@
+import org.infinispan.persistence.spi.AdvancedCacheWriter;
+import org.infinispan.persistence.spi.CacheLoader;
+import org.infinispan.persistence.spi.CacheWriter;
+import org.infinispan.persistence.spi.FlagAffectedStore;
+import org.infinispan.persistence.spi.InitializationContext;
+import org.infinispan.persistence.spi.MarshallableEntry;
+import org.infinispan.persistence.spi.MarshallableEntryFactory;
+import org.infinispan.persistence.spi.NonBlockingStore;
+import org.infinispan.persistence.spi.SegmentedAdvancedLoadWriteStore;
+import org.infinispan.persistence.spi.TransactionalCacheWriter;
+import org.infinispan.reactive.RxJavaInterop;
+import org.infinispan.util.concurrent.BlockingManager;
+import org.infinispan.util.concurrent.CompletableFutures;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+import org.reactivestreams.Publisher;
+
+import io.reactivex.rxjava3.core.Flowable;
+import io.reactivex.rxjava3.processors.FlowableProcessor;
+import io.reactivex.rxjava3.processors.UnicastProcessor;
+
+public class NonBlockingStoreAdapter<K, V> implements NonBlockingStore<K, V> {
+   private static final Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass());
+   private static final boolean trace = log.isTraceEnabled();
+   private static final AtomicInteger id = trace ? new AtomicInteger() : null;
+
+   private final Lifecycle oldStoreImpl;
+   private final Set<Characteristic> characteristics;
+
+   private BlockingManager blockingManager;
+   private MarshallableEntryFactory<K, V> marshallableEntryFactory;
+
+   public NonBlockingStoreAdapter(Lifecycle oldStoreImpl) {
+      this.oldStoreImpl = oldStoreImpl;
+      this.characteristics = determineCharacteristics(oldStoreImpl);
+   }
+
+   public Lifecycle getActualStore() {
+      return oldStoreImpl;
+   }
+
+   private String nextTraceId() {
+      return id != null ? "StoreAdapter-" + id.getAndIncrement() : null;
+   }
+
+   static private Set<Characteristic> determineCharacteristics(Object storeImpl) {
+      EnumSet<Characteristic> characteristics;
+      if (storeImpl instanceof SegmentedAdvancedLoadWriteStore) {
+          characteristics = EnumSet.of(Characteristic.SEGMENTABLE, Characteristic.EXPIRATION,
+               Characteristic.BULK_READ);
+      } else {
+         characteristics = EnumSet.noneOf(Characteristic.class);
+         if (storeImpl instanceof AdvancedCacheLoader) {
+            characteristics.add(Characteristic.BULK_READ);
+         } else if (!(storeImpl instanceof CacheLoader)) {
+            characteristics.add(Characteristic.WRITE_ONLY);
+         }
+
+         if (storeImpl instanceof AdvancedCacheWriter) {
+            characteristics.add(Characteristic.EXPIRATION);
+         } else if (!(storeImpl instanceof CacheWriter)) {
+            characteristics.add(Characteristic.READ_ONLY);
+         }
+      }
+
+      Store storeAnnotation = storeImpl.getClass().getAnnotation(Store.class);
```

Comment:
Now that we have Characteristics, I don't think there is any need for `org.infinispan.commons.persistence.Store` annotation. We should deprecate this so that we can remove it in Infinispan 14.

Reply:
👍 to deprecate. I thought about but just forgot to do so.

Reply:
https://issues.redhat.com/browse/ISPN-11865

I'll create a PR